examples/pubsub_multithread_inproc.cpp

```diff
@@ -0,0 +1,114 @@
+#include <future>
+#include <string>
+
+#include "zmq.hpp"
+
+void PublisherThread(zmq::context_t* ctx)
+{
+	//  Prepare publisher
+	zmq::socket_t publisher(*ctx, zmq::socket_type::pub);
+	publisher.bind("inproc://#1");
+
+	// Give the subscribers a chance to connect, so they don't lose any messages
+	std::this_thread::sleep_for(std::chrono::milliseconds(20));
+
+	while (true) {
+		//  Write three messages, each with an envelope and content
+		publisher.send(zmq::str_buffer("A"), zmq::send_flags::sndmore);
+		publisher.send(zmq::str_buffer("Message in A envelope"));
+		publisher.send(zmq::str_buffer("B"), zmq::send_flags::sndmore);
+		publisher.send(zmq::str_buffer("Message in B envelope"));
+		publisher.send(zmq::str_buffer("C"), zmq::send_flags::sndmore);
+		publisher.send(zmq::str_buffer("Message in C envelope"));
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+	}
+}
+
+void SubscriberThread1(zmq::context_t* ctx)
+{
+	//  Prepare subscriber
+	zmq::socket_t subscriber(*ctx, zmq::socket_type::sub);
+	subscriber.connect("inproc://#1");
+
+	//  Thread2 opens "A" and "B" envelopes
+	subscriber.setsockopt(ZMQ_SUBSCRIBE, "A", 1);
+	subscriber.setsockopt(ZMQ_SUBSCRIBE, "B", 1);
+
+	while (1) {
+		//  Read envelope address
+		zmq::message_t address;
+		zmq::recv_result_t result = subscriber.recv(address);
+
+		//  Read message contents
+		zmq::message_t contents;
+		result = subscriber.recv(contents);
+
+		std::stringstream out;
+		out << "Thread2: [" << address.to_string_view() << "] " << contents.to_string_view() << std::endl;
+		OutputDebugString(out.str().c_str());
+	}
+}
+
+void SubscriberThread2(zmq::context_t* ctx)
+{
+	//  Prepare our context and subscriber
+	zmq::socket_t subscriber(*ctx, zmq::socket_type::sub);
+	subscriber.connect("inproc://#1");
+
+	//  Thread3 opens ALL envelopes
+	subscriber.setsockopt(ZMQ_SUBSCRIBE, "", 0);
```

Comment:
The subscribes can use the type safe way `subscriber.set(zmq::sockopt::subscribe, "");`